### PR TITLE
CLI updates

### PR DESCRIPTION
**Pull request by maxinelasp**
# Change Summary
CLI updates to rework how the CLI is used

## Overview

Updated CLI to run MAG files and changed the API

# IMPORTANT NOTE: 
This is built off of the other MAG CDF changes (in another branch) and is dependent on those. Please ignore duplicate files (anything MAG specific, under the mag/ directory) from there, they will go away once that change is merged. I can't open a PR into the base repository but compare to my branch, but I want to open this to comments. 

Only files to review here: `imap_processing/cli.py`, `imap_processing/cdf/utils.py`

This is also dependent on some imap-data-access changes, up for review [here](https://github.com/IMAP-Science-Operations-Center/imap-data-access/pull/18). If you want to test this locally, set up the changes as a [git dependency](https://python-poetry.org/docs/dependency-specification/#git-dependencies) in poetry, or checkout the imap-data-access repo and set up a [file dependency](https://python-poetry.org/docs/dependency-specification/#path-dependencies).  

BOTH dependent PRs will need to be merged (and in imap-data-access, released) before this PR can be.

### CODE REVIEW

File: imap_processing/cdf/global_attrs.py
Comment (maxinelasp) [inline imap_processing/cdf/global_attrs.py]: Don't review this file

File: imap_processing/cdf/utils.py
Comment (maxinelasp) [inline imap_processing/cdf/utils.py:3]: YES review this file

File: imap_processing/cli.py
Comment (maxinelasp) [inline imap_processing/cli.py:13]: YES review this file

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [inline imap_processing/mag/l0/decom_mag.py]: Don't review this file here

File: imap_processing/mag/l0/mag_l0_data.py
Comment (maxinelasp) [inline imap_processing/mag/l0/mag_l0_data.py]: Don't review this file here

File: imap_processing/mag/l1a/mag_l1a.py
Comment (maxinelasp) [inline imap_processing/mag/l1a/mag_l1a.py:1]: Don't review this file here (though you can reference it)

File: imap_processing/mag/mag_cdf_attrs.py
Comment (maxinelasp) [inline imap_processing/mag/mag_cdf_attrs.py]: Do not review this file

File: imap_processing/mag/tests/test_mag_decom.py
Comment (maxinelasp) [inline imap_processing/mag/tests/test_mag_decom.py]: Don't review this file here

File: imap_processing/tests/cdf/test_attr_classes.py
Comment (maxinelasp) [inline imap_processing/tests/cdf/test_attr_classes.py]: Don't review this file here

File: imap_processing/cdf/utils.py
Comment (bourque) [inline imap_processing/cdf/utils.py]: I think this could also be removed since it is not used?

File: imap_processing/cli.py
Comment (bourque) [inline imap_processing/cli.py:1]: Could you also update the usage example in the module docstrings to reflect these changes?

File: imap_processing/cli.py
Comment (bourque) [inline imap_processing/cli.py]:
```suggestion
        for dependency in self.dependencies:
```

File: imap_processing/cli.py
Comment (bourque) [inline imap_processing/cli.py]:
```suggestion
                value = imap_data_access.query(
```

File: imap_processing/cli.py
Comment (bourque) [inline imap_processing/cli.py:307]: Larger philosophical question outside the scope of this PR: since we are really now starting to implement these, should we move these classes to a separate module? Either one module per instrument, or one module that stores all of the instruments?

File: imap_processing/cli.py
Comment (bourque) [inline imap_processing/cli.py]: It might be better if we name this with the word "upload" somewhere in it, since that is what it is used for? Maybe `--upload-to-cloud` or `--upload-to-aws` or `--upload-to-sdc`?

File: imap_processing/cli.py
Comment (bourque) [reply imap_processing/cli.py:307]: I fear that if we leave this the way it is, this module is going to get very large!

File: imap_processing/cdf/utils.py
Comment (maxinelasp) [reply imap_processing/cdf/utils.py]: Merge fail 😆

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py]: `--api-upload`?

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py:307]: Yeah good question! Yeah, I think it would make sense to move the ProcessInstrument class and all the subclasses into a different file and keep the CLI stuff here

File: imap_processing/cdf/utils.py
Comment (tech3371) [inline imap_processing/cdf/utils.py:53]: Do we assume here that user had already constructed path using `imap-data-access` library which validates filename?

File: imap_processing/cli.py
Comment (tech3371) [reply imap_processing/cli.py]: I like this one `--upload-to-sdc`. It kind of tells what this flag do.

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py:347]: Yes. I can update this afterwards.

File: imap_processing/mag/__init__.py
Comment (tech3371) [inline imap_processing/mag/__init__.py]: I don't know if you use this version for filename. If you do, this is not in the format it expects and will throw error.

File: imap_processing/cdf/utils.py
Comment (maxinelasp) [reply imap_processing/cdf/utils.py:53]: Yes, perhaps I should add that into the comment. This assumes that the path is correct, or at least that you are sure it is the path you want to write the file to. I think this method should not be responsible for validating that - for example, if you want to write test files locally that don't match our naming convention, this method should allow for that.

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py:347]: Cool, I figured I could leave more of the SWE stuff for you to finalize, but I just didn't want to break it in my PR 😆

File: imap_processing/mag/__init__.py
Comment (maxinelasp) [reply imap_processing/mag/__init__.py]: No, this is the version for the instrument code, which I believe only gets written into the CDF files. The CLI gets a version passed in, and does not generate versions at all.

File: imap_processing/cdf/utils.py
Comment (tech3371) [reply imap_processing/cdf/utils.py:53]: ok

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: can we update example to match new values passed in?
```
[
        {
            'instrument': 'mag',
            'data_level': 'l0',
            'descriptor': 'sci',
            'version': 'v00-01',
            'start_date': '20231212',
            'end_date': '20231212'
        }]
   ```

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: same here. some doc string update

File: imap_processing/cli.py
Comment (tech3371) [reply imap_processing/cli.py:347]: I think you did great job on that. I saw SWE processed successfully last week.

File: imap_processing/mag/__init__.py
Comment (tech3371) [reply imap_processing/mag/__init__.py]: cool.

**Review by tech3371 — COMMENTED** [review]
Couple minor doc string updates. Otherwise, it looks good to me!

File: imap_processing/codice/utils.py
Comment (tech3371) [inline imap_processing/codice/utils.py]: I think this is a topic of discussion. CDF wants one thing but I think instrument team may still want raw spacecraft time.

File: pyproject.toml
Comment (tech3371) [inline pyproject.toml]: Did we decide to update this to latest release Bryan made this week which is `1.2.4` or wait? Sorry, I don't remember what we decided.

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: I think we are safe to delete this argument now.

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]:
```suggestion
        help="Start time for the output data. Format: YYYYMMDD",
```

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]:
```suggestion
        "for end_time. Format: YYYYMMDD",
```

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]:
```suggestion
        help="Version of the data. Format: vxx-xx",
```

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: can we update doc strings to match new inputs?

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: was there reason for using hardcoded version? should we use `self.version` now?

File: imap_processing/cli.py
Comment (tech3371) [inline imap_processing/cli.py]: a dump question, how does say `--start-date` gets converted to `args.start_date`? I looked but I am not seeing it where this conversion happens.

File: imap_processing/codice/utils.py
Comment (maxinelasp) [reply imap_processing/codice/utils.py]: Well for now, the tests fail if there isn't datetime64, so I think regardless we will want to convert from SHCOARSE into datetime64. Is it ok if I leave this for now?

File: pyproject.toml
Comment (maxinelasp) [reply pyproject.toml]: This poetry.lock file does upgrade it, but it should still work with `1.2.3` so I left it like this. I can force a version bump if that makes sense though

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py]: Good catch!

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py]: Oh yeah, you're totally right. Good catch!

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py]: You know, I left it this way with the intent to go back to it, and then it all worked so I never did. Looks like `args` automatically converts the arguments from a hyphen to an underscore.

File: imap_processing/codice/utils.py
Comment (tech3371) [reply imap_processing/codice/utils.py]: yes. We can address that in upcoming weeks.

**Review by tech3371 — APPROVED** [review]
Once you test that this works in AWS, it looks good to me! And we can address other comments in upcoming PR.

**Review by bourque — APPROVED** [review]
Lots of work went into this PR, nice job!

File: imap_processing/codice/utils.py
Comment (bryan-harter) [reply imap_processing/codice/utils.py]: You can have multiple time variables in your CDF file, so there can be another variable for spacecraft clock time.  

But the one variable named "epoch" must be of the type "CDF_TT2000".  And all DEPEND_0s need to point to the "epoch" variable.

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py:16]: I find "loads" to be pretty generic and not knowing where it comes from is confusing. I'd slightly prefer this to be: `import json` and then use it as `json.loads()` where needed.

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py]:
```suggestion
        '   "start_date": "20231212"',
```

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py]:
```suggestion
        help="Upload completed output files to the IMAP SDC.",
```

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py:203]: Are we dumping malformed json somewhere upstream? If so we should correct it there. Single quotes aren't allowed in the JSON spec, so I'm not sure we should be doing the replacement beforehand.

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py:209]: warn/log instead of printing? There is no way to avoid this print otherwise.

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py:228]: This would mean I can't run local processing without an internet connection/access to the SDC servers? Do we need to add some checks in the query to add a "local" command to query the local files too? This can come in follow-up issues/PRs

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py:244]: Above you have the "Unable to download files" message. Should this be moved up into that try/except block so we also catch the HTTPError there?

How is return_query sorted? I would guess we want the latest file `[-1]` or all of them even by iterating over the list returned?

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py]: self.download_dependencies()? Or just remove the comments for now if not needed.

File: imap_processing/mag/l1a/mag_l1a.py
Comment (greglucas) [inline imap_processing/mag/l1a/mag_l1a.py:17]: cdf_filepath is not super descriptive to me of what that is referring to (cdf inputs, outputs, IMAP base directory), if this is where you want the output to go I'd suggest `output_filepath` as a more explicit naming. But, not blocking at all on this.

File: imap_processing/tests/idex/test_l1_cdfs.py
Comment (greglucas) [inline imap_processing/tests/idex/test_l1_cdfs.py]:
```suggestion
@pytest.mark.xfail(reason="Need to fix new ISTP error in IDEX CDF")
```

File: imap_processing/tests/idex/test_l1_cdfs.py
Comment (greglucas) [inline imap_processing/tests/idex/test_l1_cdfs.py]: In the new implementation of the `write_cdf()` are we expanding the `input_name` to the imap-data-access data location? (Which in the test suite we set to a temporary directory)

I'm concerned that these are going to be written to the local directories and start adding files we don't want while running the tests. This is skipped currently, but taking off the skip locally a file was produced there.

**Review by greglucas — CHANGES_REQUESTED** [review]
I am confused where data is supposed to live and just want that clarified before merging since there are a lot of binary files in various locations now.

We have a new `data/imap/mag` folder nested under testing, which I am not sure should be in the tests, or is it a test artifact that only needs to be produced and thus should be in temporary directories?

https://github.com/IMAP-Science-Operations-Center/imap_processing/tree/dev/imap_processing/data/imap/mag

We also have this file now too: https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/dev/imap_processing/tests/imap_swe_l1_test_20100101_20100101_v01.cdf

I guess this is just a concern on my part that we will accidentally be doing `git commit .` and adding unnecessary/unexpected files.

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py:244]: Right now, we would expect only one file to return for a given dependency query. That's not the case right now, but Tenzin needs to update the dependency management in batch_starter first.

File: imap_processing/tests/idex/test_l1_cdfs.py
Comment (maxinelasp) [reply imap_processing/tests/idex/test_l1_cdfs.py]: Right, I'll update the input_name path to write to a temp directory, I didn't catch that. I'm also open to adding back in the `directory` keyword to `write_cdf`

**Comment by maxinelasp** [comment]
I'll go in and check the tests to see if they're producing files outside of a temp directory. The `imap/mag` folder is just a fail from me adding it, I think, since that structure is from `imap-data-access`.

File: imap_processing/cli.py
Comment (maxinelasp) [reply imap_processing/cli.py:203]: I'm not sure, this piece is left over from the first draft (just moved). If we want to change the JSON that would be in batch_starter.

Comment (tech3371) [reply imap_processing/cli.py:203]: This is what get's passed to batch job(aka here). I took this from most recent logs
```
Submitting job with this command - ['--instrument', 'mag', '--data-level', 'l1b', '--start-date', '20231212', '--end-date', '20231212', '--version', 'v00-01', '--dependency', "[{'instrument': 'mag', 'data_level': 'l1a', 'version': 'v00-01'}]", '--upload-to-sdc']
```

File: imap_processing/cli.py
Comment (tech3371) [reply imap_processing/cli.py:244]: To expand more on what Maxine said, before Maxine had above query like this
```
                return_query = imap_data_access.query(
                    instrument=dependency["instrument"],
                    data_level=dependency["data_level"],
                    version=dependency["version"],
                    start_date=dependency["start_date"],
                    end_date=dependency["end_date"],
                    descriptor=dependency["descriptor"],
                )
```
I asked her to temporarily changed that to this for now
```
                return_query = imap_data_access.query(
                    instrument=dependency["instrument"],
                    data_level=dependency["data_level"],
                    version=dependency["version"],
                )
```
The reason was because I don’t have all the information in the dependency list yet. That’s not done yet. Right now, below is what's passed to batch job, as I mentioned above. This is one of the TODO in the big action items list post demo ticket. I haven’t got around to it because that involves many changes such as database changes and make sure we didn’t mess up dependency checks. I know I am not gonna be able to make that change in time if I was to do it.
```
['--instrument', 'mag', '--data-level', 'l1b', '--start-date', '20231212', '--end-date', '20231212', '--version', 'v00-01', '--dependency', "[{'instrument': 'mag', 'data_level': 'l1a', 'version': 'v00-01'}]", '--upload-to-sdc']
```

**Comment by greglucas** [comment]
> I'll go in and check the tests to see if they're producing files outside of a temp directory. The imap/mag folder is just a fail from me adding it, I think, since that structure is from imap-data-access.

Thanks! I think this was my primary concern and reason for not wanting to allow individual output locations at first, just so we could keep things a little more isolated somehow. I'm not sure what the best approach is here though because I don't think we'd want both the directory and path allowed as they might conflict if you pass in a full path. Is there a way to make the filepath argument optional and deduce the path structure from the xarray dataset passed in? So default to no path and then write everything to the imap-data-access location (temporary within the tests), but allow users to specify a path if they know what they are doing and want to push it somewhere else.

File: imap_processing/cli.py
Comment (greglucas) [reply imap_processing/cli.py:244]: Sounds good! Just wanted to make sure this had a reason, and both of you are on it which is great :)

File: imap_processing/cli.py
Comment (greglucas) [reply imap_processing/cli.py:203]: @tech3371 where does that command get created? I think it just needs a `json.dumps()` surrounding the dependency argument which will invert that string escaping for us.

Comment (tech3371) [reply imap_processing/cli.py:203]: here: https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/dev/sds_data_manager/lambda_code/SDSCode/batch_starter.py#L315

**Comment by maxinelasp** [comment]
> Is there a way to make the filepath argument optional and deduce the path structure from the xarray dataset passed in? 

I think this is a good idea... but do you want to block the PR on this? Ultimately I think the answer might be to pass in a `ScienceFilePath` instance or something, but I'm not sure. Ultimately, that question isn't super related to the CLI changes, and I don't think it's worth trying to fix here, since this PR has such an impact on sds-data-manager. 

I did remove the files you were concerned about (thanks for catching that!) and update the tests that generate CDFs - can you take another look?

**Review by greglucas — APPROVED** [review]

**Comment by greglucas** [comment]
> I did remove the files you were concerned about (thanks for catching that!) and update the tests that generate CDFs - can you take another look?

Yep, this was what I was blocking over, so thanks for fixing it! Merge away before you take off :)

**Pull request merged**
